components/wifi/esp8266-driver/ESP8266Interface.cpp

```diff
@@ -194,11 +232,33 @@ int ESP8266Interface::connect()
         return NSAPI_ERROR_DHCP_FAILURE;
     }
 
-    return _esp.connect(ap_ssid, ap_pass);
+    _cmutex.lock();
+
+    MBED_ASSERT(!_connect_event_id);
+    _connect_event_id = _global_event_queue->call(callback(this, &ESP8266Interface::_connect_async));
+
+    if (!_connect_event_id) {
+        MBED_ERROR(MBED_MAKE_ERROR(MBED_MODULE_DRIVER, MBED_ERROR_CODE_ENOMEM), \
+                        "connect(): unable to add event to queue");
+    }
+
+    if (_if_blocking) {
+        _if_connected.wait();
```

Comment:
This needs to be a proper condition test loop. Reread the CV documentation

    while (connection_status != CONNECTED) {
        _if_connected.wait();
    }